pandas/core/dtypes/cast.py

```diff
@@ -336,74 +353,427 @@ def changeit():
         pass
-    elif issubclass(np.dtype(dtype).type, str):
-        dtype = np.object_
+    elif isna(fill_value) or _is_iNaT(fill_value):
+        # cast missing values (incl. iNaT) to correct missing value marker for
+        # the updated dtype
+        fill_value = na_value
+    # otherwise cast fill_value (= only entry of fill_array) to new dtype
+    elif is_datetime_or_timedelta_dtype(dtype):
+        # for datetime/timedelta, we need to return the underlying ints
+        fill_value = fill_array.astype(dtype)[0].value
+    else:
+        fill_value = fill_array.astype(dtype)[0]
 
     return dtype, fill_value
 
 
+def _maybe_promote_with_array(dtype, fill_value=np.nan):
+    """
+    Determine minimal dtype to hold fill_value, when starting from dtype
+
+    This will also return the default missing value for the resulting dtype, if
+    necessary (e.g. for datetime / timedelta, the missing value will be `iNaT`)
+
+    Parameters
+    ----------
+    dtype : DType
+        The dtype to start from.
+    fill_value : np.ndarray / Series / Index
+        Array-like of values that the output dtype needs to be able to hold.
+
+    Returns
+    -------
+    dtype : DType
+        The updated dtype.
+    na_value : scalar
+        The missing value for the new dtype. Returns None or dtypes that
+        cannot hold missing values (integers, booleans, bytes).
+
+    See Also
+    --------
+    _maybe_promote_with_scalar : similar method for scalar case
+
+    Examples
+    --------
+    >>> maybe_promote(np.dtype('int'), fill_value=np.array([None]))
+    (dtype('float64'), nan)
+    >>> maybe_promote(np.dtype('float'), fill_value=np.array(['abcd']))
+    (dtype('O'), nan)
+
+    For datetimes, timedeltas and datetimes with a timezone, the missing value
+    marker is pandas._libs.tslibs.iNaT (== np.iinfo('int64').min):
+
+    >>> maybe_promote(np.dtype('datetime64[ns]'), fill_value=np.array([None]))
+    (dtype('<M8[ns]'), -9223372036854775808)
+
+    String values do not get cast to datetime/timedelta automatically, but
+    force an upcast to object (with corresponding missing value marker nan).
+
+    >>> maybe_promote(np.dtype('datetime64[ns]'),
+    ...               fill_value=np.array(['2018-01-01']))
+    (dtype('O'), nan)
+
+    The method will infer as conservatively as possible for integer types:
+
+    >>> maybe_promote(np.dtype('uint8'),
+    ...               fill_value=np.array([np.iinfo('uint8').max + 1]))
+    (dtype('uint16'), None)
+    >>> maybe_promote(np.dtype('uint8'), fill_value=np.array([-1]))
+    (dtype('int16'), None)
+    """
+
+    if isinstance(fill_value, np.ndarray):
+        if fill_value.ndim == 0:
+            # zero-dimensional arrays cannot be iterated over
+            fill_value = np.expand_dims(fill_value, 0)
+        elif fill_value.ndim > 1:
+            # ndarray, but too high-dimensional
+            fill_value = fill_value.ravel()
+    elif not isinstance(fill_value, (ABCSeries, ABCIndexClass)):
```

Comment:
do we need Series/Index?  ATM we just have ndarray right?

Reply:
I see no harm in permitting them (*if* arrays are permitted at all), as they they fit into the code without extra effort (and later uses of `maybe_upcast_putmask` might well plop a Series in there).

Reply:
Until we have a compelling use case, let's restrict the inputs to 1D, non-empty ndarray

Reply:
OK